src/NUnitFramework/framework/Internal/TestNameGenerator.cs

```diff
@@ -583,6 +588,52 @@ public override string GetText(MethodInfo method, object[] args)
             }
         }
 
+        private class ParamArgListFragment : NameFragment
+        {
+            private readonly int _maxStringLength;
+
+            public ParamArgListFragment(int maxStringLength)
+            {
+                _maxStringLength = maxStringLength;
+            }
+
+            public override string GetText(MethodInfo method, object[] args)
+            {
+                var sb = new StringBuilder();
+                args = args ?? new object[0];
+
+                var parameters = method.GetParameters();
+
+                sb.Append('(');
+                var count = Math.Max(args.Length, parameters.Length);
+
+                for (int i = 0; i < count; i++)
+                {
+                    if(i < parameters.Length)
+                    {
+                        if (parameters[i].IsOptional)
+                        {
+                            if(i >= args.Length)
+                            {
+                                continue;
```

Comment:
Not `break`?
Also, the if statements can be joined with `&&`.

Reply:
I was thinking over a 'break'.
If possible, specify a non-optional parameter between optional, this is not suitable.
At the moment there is no such possibility, perhaps it never will be.
I can change to break.

Reply:
>the if statements can be joined with &&.
I prefer brackets, I can also change.

Reply:
That's a good thing to consider, but I think the `break;` also pairs well with the `i > args.Length`.

I'd be happy if we disregarded `IsOptional`, even, for this purpose, unless you think that could be a source of confusion.

Reply:
If an optional parameter is not specified, then its default value can be displayed.
For tests, I never used the optional parameters, so I do not have a clear opinion.

Reply:
Perhaps I did not quite understand your point.

Reply:
By disregarding IsOptional, I meant, let's pretend for the sake of formatting that all the parameters are optional if `args` is too short.

Reply:
Test names, IMO, should reflect what the user specified foe arguments. Unspecified arts shouldn't appear. Consequently, as @jnm2 points out, it doesn't matter if they are optional.

Reply:
https://user-images.githubusercontent.com/9452368/66336959-d4484780-e946-11e9-9e65-b17b03be715a.png

In the current implementation, all arguments will be displayed as `?` Unfortunately.

Reply:
It's a non-obvious problem, but I'd hazard a guess that we've spent more effort on invalid test display over the years than valid ones! In early days, invalid tests just disappeared, which gives you a false positive. In this case, the ? Gives a warning so it's not terrible, but it is different from the default display. @jnm2 what do you think?

Reply:
@CharliePoole 
@jnm2 
In the latest version.
1. If the optional parameter is not set, it is not displayed.
2. If the required parameter is not specified, "?" is displayed instead of value.
3. If there are too many arguments, then the names of non-existent parameters are replaced with "?"

Which of these do you not like and what to change?

Reply:
In the case of {a}

1. If an optional parameter is not set, it is not displayed.

2. If a required parameter is not set, it is not displayed.

3. If there are too many arguments, they are displayed.

If we want {p} error handling to conform closely to that for {a}, then I'd change 2 and 3. In the case of 3, I'd fall back on merely displaying the value.

As usual, I'm afraid it takes seeing an example to enable me to zero in on what I'd like to see.

Reply:
What about
```csharp
[TestCase("{m}{p}", new [] { 1 }, Ex.Result = "Test.Args(a: 1)")]
[TestCase("{m}{p}", new [] { 1, 2 }, Ex.Result = "Test.Args(a: 1, b: 2)")]
[TestCase("{m}{p}", new [] { 1, 2, 3 }, Ex.Result = "Test.Args(a: 1, b: 2, c: 3)")]
[TestCase("{m}{p}", new [] { 1, 2, 3, 4 }, Ex.Result = "Test.Args(a: 1, b: 2, c: 3, ?: 4)")]

private void TestMethodWithArgs(Int32 a, Int32 b, Int32 c = 0) { }
```

Reply:
or

[TestCase("{m}{p}", new object[] { 1 }, ExpectedResult = "TestMethodWithArgs(a: 1)")]
        [TestCase("{m}{p}", new object[] { 1, 2 }, ExpectedResult = "TestMethodWithArgs(a: 1, b: 2)")]
        [TestCase("{m}{p}", new object[] { 1, 2, 3 }, ExpectedResult = "TestMethodWithArgs(a: 1, b: 2, c: 3)")]
        [TestCase("{m}{p}", new object[] { 1, 2, 3, 4 }, ExpectedResult = "TestMethodWithArgs(a: 1, b: 2, c: 3, 4)")]

Reply:
Personally, I prefer the version without the question mark.

The issue I see is that the naming of a test has never depended on the number of args that the method takes. With {a} it is solely dependent on what args the user supplies, whether they are correct, too few or too many. With {p} we obviously have to look at the names of the method parameters but I'd rather not see us change anything more than is needed. I also find it a bit surprising to see `?` used for both the name and the value.

@jnm2 Do you have any opinion one way or the other here?

Reply:
@CharliePoole Yes, my opinion is the same as yours.